### ISSUE TITLE
Suggestion about exceptions

### ISSUE THREAD

**Issue by The-Futurist** [dotnet/runtime#51863]
This a suggestion to get some idea of how others see this, if there's a better repo for this please do let me know.

Very simply it would be a huge help if the exception base class include a DateTime property that contains the date and time that the exception was thrown by the runtime.

There's no easy way to do this as a consumer of .Net because we'd need to introduce our own new base class (derived from Exception) and then derive all of our exceptions from that, additionally there's no mechanism, for an exception object to receive a method call when the throw itself executes. Yes one could put the setting of the property in the constructor but that isn't reliable as some (unusual designs) throw pre-created exceptions that they pull from lookup tables sometimes and so on.

A common and time consuming problem in production support environments is unexpected exceptions being thrown in code that often may not (for all kinds of reasons including legacy logic or just plain human oversight) and we get sent a stack trace or screenshot of a stacktrace and have to then chase people for date time info, if we get the wrong or misleading date time info it can cost us hours in wasted time.

If every exception contained a public get property that was auto populated when the throw statement was executed this problem would almost vanish, the human element of determining the date time thrown would vanish and greatly improve our ability to respond helpfully to production issues, some of which impact customers and require rapid problem solving ability.

If we - in a support role - get sent a screenshot for example that contains a stack trace (which is common in fragile ASP.NET web apps and so on) the trace info never reveals anything about the date and time the event occurred this means its very hard to associate the exception with any other event information like possible related log messages, we can't ask questions like "did this occur before or after we did X?" or "was anything else at all logged by any other app at around the same time?" and so on.

If the "Exception details" text that's present in the stack trace was then augmented to include the date-time this would be always visible to support staff.

e.g., this is a typical stack trace we might see (this is taken from the web, not from our systems):

![image](https://user-images.githubusercontent.com/12262952/116100785-1f799680-a662-11eb-83c5-5185ac251a5d.png)

**Comment by KalleOlaviNiemitalo** [comment]
Perhaps you could handle the [AppDomain.FirstChanceException](https://docs.microsoft.com/dotnet/api/system.appdomain.firstchanceexception?view=net-5.0) event and store the time stamp in [Exception.Data](https://docs.microsoft.com/dotnet/api/system.exception.data?view=net-5.0). Although I feel uneasy about how to ensure that the event handler won't ever throw another exception and cause recursion.

**Comment by CyrusNajmabadi** [comment]
> If the "Exception details" text that's present in the stack trace was then augmented to include the date-time this would be always visible to support staff.

This seems to be the crux of the request. In which case, I think this request sound go to asp.net. That said, I think it would not be likely to change that legacy exception handler. Instead, you could swap it that exception handler with your own, which also prints out the datetime at the crash point.

**Comment by Korporal** [comment]
> > If the "Exception details" text that's present in the stack trace was then augmented to include the date-time this would be always visible to support staff.
> 
> This seems to be the crux of the request. In which case, I think this request sound go to asp.net. That said, I think it would not be likely to change that legacy exception handler. Instead, you could swap it that exception handler with your own, which also prints out the datetime at the crash point.

The ASP.NET example was presented as a scenario that comes up fairly often it is not the only use case for this. Catching the timestamp of the throw records the datetime at the point closest to the event. Of course when the exception is caught the timestamp can be created or it can be inferred by looking at a log if the caught exception is logged, in an ideal world this would be enough.

But unfortunately many real world systems are far from well engineered and so relying on sound engineering practices for resolving problem is a risk, investigating serious operational problems in these cases can be time consuming and often relies on evidence gathering before creating hypotheses to explain what happened.

Additionally logging systems are distinct from the app, they may reside or execute in servers in different locations and time zones and reflect the local time the log request was received unless the originating app intentionally gets and includes the time stamp itself - which becomes unnecessary if every exception always includes the local timestamp pertaining to the throw.

I recognize that many of these odd problems can in principle be avoided by sound design and engineering principles but unfortunately many real world systems fail to meet such standards for all kinds of reasons.

Being assured that every exception always contains the timestamp reflecting when it was thrown locally means we can possibly resolve problems sooner with less information and less reliance on other people, systems or departments.

**Comment by CyrusNajmabadi** [comment]
If you have these as different servers in different timezones, you'll need to translate the info from one server to another. So you still have to do the work to do the timezone management. 

It's there a reason you have time zones here at all?  Wouldn't it make me sense for all your logs to always use utc?

**Comment by Korporal** [comment]
> If you have these as different servers in different timezones, you'll need to translate the info from one server to another. So you still have to do the work to do the timezone management.
> 
> It's there a reason you have time zones here at all? Wouldn't it make me sense for all your logs to always use utc?

Having the exception timestamp inside the exception means one can unambiguously log that as a datum no matter what logging systems one has or what time zone a user may be in. Resolving operational problems frequently involves answering the question "OK when did the user do this?" or "OK when did this actually take place?" or "Did this arise before or after the evening batch suite started?" and so on.

Also we are concerned with knowing when an exception was thrown not when a logging system received a message and wrote it to the log, in most cases these will be milliseconds apart but there is a window there and in busy systems with huge workloads one can envisage delays perhaps on the seconds or more, in fact without knowing the exception thrown timestamp one could never even detect if there had been a delay in the logging of such an event.

**Comment by CyrusNajmabadi** [comment]
> Also we are concerned with knowing when an exception was thrown not when a logging system received a message and wrote it to the log, in most cases these will be milliseconds apart but there is a window there and in busy systems with huge workloads one can envisage delays perhaps on the seconds or more

That is true.  But in my own experience hasn't been an issue.  Furthermore, if milliseconds apart is an issue, you fundamentally are not going to find that this works as your clocks in different regions for all those servers you have are also not going to agree on this stuff.  

Given the stack tells you where the exception happened, can you just solve this with a try/catch there that logs and rethrows.  This will give you the timestamp you want, and will allow you to collect any and all other information you need as close to the source to determine what the issue is.

**Comment by Korporal** [comment]
> > Also we are concerned with knowing when an exception was thrown not when a logging system received a message and wrote it to the log, in most cases these will be milliseconds apart but there is a window there and in busy systems with huge workloads one can envisage delays perhaps on the seconds or more
> 
> That is true. But in my own experience hasn't been an issue. Furthermore, if milliseconds apart is an issue, you fundamentally are not going to find that this works as your clocks in different regions for all those servers you have are also not going to agree on this stuff.
> 
> Given the stack tells you where the exception happened, can you just solve this with a try/catch there that logs and rethrows. This will give you the timestamp you want, and will allow you to collect any and all other information you need as close to the source to determine what the issue is.

I don't think delays of milliseconds are a concern and as you say if they are then we're dealing with a different problem altogether. In reality there can be intervals in the _thousands_ of milliseconds (one, two etc seconds) and this in turn leads to the next concern that the ordering of logged events too can be vital to correct problem diagnosis, in a system where there is a variable and unpredictable delay in logging we will see messages ordered in time of arrival not necessarily in order of time thrown.

The question we are ultimately concerned about is when did this occur and not when was this logged, its preferable to eliminate details like logging delays, log server clock settings and time zones completely from the event being logged, these are superfluous and really just a means to end which is - when did this occur and what is the true temporal order of these logged exceptions, recording the throw timestamp at the sources seems to be the most sensible starting point.
